core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java

```diff
@@ -560,6 +592,252 @@ public UpdateNode plan(Update node)
                 outputs);
     }
 
+    private MergeCaseKind getMergeCaseKind(MergeCase mergeCase)
+    {
+        requireNonNull(mergeCase, "mergeCase is null");
+        if (mergeCase instanceof MergeInsert) {
+            return MergeCaseKind.INSERT;
+        }
+        if (mergeCase instanceof MergeUpdate) {
+            return MergeCaseKind.UPDATE;
+        }
+        if (mergeCase instanceof MergeDelete) {
+            return MergeCaseKind.DELETE;
+        }
+        throw new IllegalArgumentException("Unrecognized MergeCase " + mergeCase.getClass());
+    }
+
+    public MergeNode plan(Merge merge)
+    {
+        MergeAnalysis mergeAnalysis = analysis.getMergeAnalysis().orElseThrow(() -> new IllegalArgumentException("analysis.getMergeAnalysis() isn't present"));
+
+        List<ColumnHandle> redistributionColumns = mergeAnalysis.getRedistributionColumnHandles();
+        Map<Integer, List<ColumnHandle>> mergeCaseColumnsHandles = mergeAnalysis.getMergeCaseColumnHandles();
+
+        MergeDetails mergeDetails = createMergeDetails(merge, mergeCaseColumnsHandles);
+
+        RelationPlan target = new RelationPlanner(analysis, symbolAllocator, idAllocator, lambdaDeclarationToSymbolMap, metadata, outerContext, session, recursiveSubqueries)
+                .process(merge.getTable());
+
+        PlanBuilder targetPlanBuilder = newPlanBuilder(target, analysis, lambdaDeclarationToSymbolMap);
+
+        Assignments.Builder projections = Assignments.builder();
+        projections.putIdentities(targetPlanBuilder.getRoot().getOutputSymbols());
+
+        Symbol presentColumn = symbolAllocator.newSymbol("present", BOOLEAN);
+        projections.put(presentColumn, TRUE_LITERAL);
+
+        target = new RelationPlan(
+                new ProjectNode(idAllocator.getNextId(), targetPlanBuilder.getRoot(), projections.build()),
+                analysis.getScope(merge.getTable()),
+                target.getFieldMappings(),
+                outerContext);
+
+        RelationPlan source = new RelationPlanner(analysis, symbolAllocator, idAllocator, lambdaDeclarationToSymbolMap, metadata, outerContext, session, recursiveSubqueries)
+                .process(merge.getSource());
+
+        RelationPlan joinPlan = new RelationPlanner(analysis, symbolAllocator, idAllocator, lambdaDeclarationToSymbolMap, metadata, outerContext, session, recursiveSubqueries)
+                .planJoin(merge.getPredicate(), Join.Type.RIGHT, mergeAnalysis.getJoinScope(), target, source, analysis.getSubqueries(merge));
```

Comment:
> Please note that there are no subqueries recorded for merge.

I'm not sure what this means.  What code should be changed, and how?

Reply:
In `StatementAnalyzer`, you do `recordSubqueries()` three times:
```
analysis.recordSubqueries(mergePredicate, predicateAnalysis);
```
and
```
analysis.recordSubqueries(operation, analyzeExpression(predicate, joinScope));
```
and
```
analysis.recordSubqueries(operation, expressionAnalysis);
```
The first argument of those calls is the map key. So these are recorded under `mergePredicate` or `operation` (which is a MergeCase).

The call in `QueryPlanner` that I'm mentioning in my comment: `analysis.getSubqueries(merge)`, retrieves values from that map under the key `merge`. The result is empty, because nothing was ever recorded under that key. 

My suggestion is to record all subqueries under the enclosing `Merge` node, and make sure you access them by the right key.

Reply:
> The result is empty, because nothing was ever recorded under that key.

Do we have tests with subqueries? 
I understand we do not have them yet, if we identified a problem, but the build is green.

Reply:
@kasiafi:  We'll follow up in person on your comment.

@findepi:
> Do we have tests with subqueries?

There are a bunch of tests with subqueries, but maybe not in suitable places.  Several tests run merges with this shape:

```
        query(format("MERGE INTO %s t USING (SELECT * FROM (VALUES %s)) AS s(customer, purchases, zipcode, spouse, address)", targetTable, firstMergeSource) +
                "    ON t.customer = s.customer" +
                "    WHEN MATCHED THEN UPDATE SET purchases = s.purchases, zipcode = s.zipcode, spouse = s.spouse, address = s.address");
```

Here is a merge that has a subquery in the predicate:

```
                query(format("MERGE INTO %s t USING %s s", targetTable, sourceTable) +
                        "    ON (t.nation_name = s.nation_name)" +
                        "    WHEN MATCHED AND t.nation_name > (SELECT name FROM tpch.tiny.region WHERE name = t.region_name AND name LIKE ('A%'))" +
                        "        THEN DELETE" +
                        "    WHEN NOT MATCHED AND s.region_name = 'EUROPE'" +
                        "        THEN INSERT VALUES(s.nation_name, s.region_name)");
```

Perhaps these don't cover the interesting cases.  I'm always happy to write more tests; please suggest some cases to be covered.

Reply:
@kasiafi and I discussed this.  I need to add tests of a subquery in the search expression (above `t.nation_name = s.nation_name`), and in the column values in `INSERT` and `UPDATE` merge cases.

Reply:
@findepi - - as part of reworking handling of subqueries in `MERGE`, I improved the `MERGE` subquery test to have a subquery in every expression in the `MERGE` tree: the `MERGE` search condition; the `MERGE` case predicate; and the `setExpression`s that result from `INSERT` and `UPDATE` columns.